### ISSUE TITLE
Bring MaxUnavailable Rolling Update to StatefulSet

### ISSUE THREAD

**Issue by krmayankk** [kubernetes/kubernetes#68397]
**Is this a BUG REPORT or FEATURE REQUEST?**:
/kind feature

 I would like to add maxUnavailable to StatefulSet  and i am seeking feedback and use cases from the community.

Many times we want stable network identity(StatefulSet) but would like to be able to do a rolling update on it without going one pod at a time . This helps in faster rollout of changes. Examples of such scenarios are:-

1:  My containers publish metrics to time series system. If I am using a Deployment , each rolling update creates a new pod name and hence the metrics published by these new pod names , starts a new time series which makes tracking metrics for the application difficult. While this could be mitigated, it requires some tricks on the time series collection side. It would be so much nicer if i could use a StatefulSet object so my object names doesnt change and hence all metrics goes to a single time series, but still get all the other benefits of a Deployment like maxUnavailable

2: My Container do some initial startup tasks like loading up cache or something that takes a lot of time. If we used StatefulSet, we can only go one pod at a time which would result in a slow rolling update. If we did maxUnavailable for StatefulSet with a greater than 1 number, it would allow for a faster rollout. THe bursting using parallel pod management does allow some improvement, but it doesnt control how many to burst, so it can literally bring down the entire application.

3: Even for a truly stateful clustered application, with a cluster of n nodes having the same role(follower, readers , etc), if my application can tolerate it i want to be able to faster rollouts by bringing down 2 or more at a time.

4: Sometimes i just want easier tracking of revisions of a rolling update. Deployment does it through ReplicaSets and has its own nuances. Understanding that requires diving into the complicacy of hashing and how replicasets are named. Over and above that, there are some issues with hash collisions which further complicate the situation(I know they were solved). StatefulSet introduced ControllerRevisions in 1.7 which I believe are easier to think and reason about. They are used by DaemonSet and StatefulSet for tracking revisions. It would be so much nicer if all the use cases of Deployments can be meant and we could track the revisions by ControllerRevisions.

Are there other uses cases for this , people are looking for ?

Basically i feel like the lines have completely blurred between StatefulSet and Deployment and if there were things available in StatefulSet that are already available in Deployment, like maxUnavailable, I don't see many strong reasons to use a Deployment (other than obviously it being more battle tested and stable then StatefulSet)

Thoughts from the community ?

Do you support adding maxUnavailable to StatefulSet ?

**Comment by krmayankk** [comment]
@kow3ns  do you agree with this work ? Do you see other people asking for this ?

**Comment by janetkuo** [comment]
We discussed this in SIG Apps meeting today and we're looking for more requests from users and some more concrete use cases. 

I'd also like to point out that MaxSurge does not make sense for StatefulSets -- you wouldn't want to create more replicas (surge) than a StatefulSet's `.spec.replicas` during a rolling update, as each pod has its own identity. MaxUnavailable is the only thing needed here.

**Comment by krmayankk** [comment]
thanks @janetkuo  makes sense. I am going to float this idea on k8s-dev and k8s-users  to see what others think about this

**Comment by mattfarina** [comment]
We discussed this in SIG Apps. We are OK taking maxUnavailable to the next step. We discussed how you can do this today and how maxUnavailable could be a better experience.

The next step for this would be a [KEP](https://github.com/kubernetes/community/tree/master/keps) for the features. There is [a template](https://github.com/kubernetes/community/blob/master/keps/0000-kep-template.md) that can be used as a starting point.

This KEP should list the owning SIG as `sig-app` and can go in that sub-directory.

Using a KEP follows the current processes and since this does make an API change (a backwards compatible addition) it will help facilitate going through an API review.

Is there someone who wants to craft the KEP for this?

**Comment by krmayankk** [comment]
@mattfarina I will go ahead and Start Writing this KEP. Thanks for discussing this in Sig Apps and bringing it forward.

**Comment by krmayankk** [comment]
I have a rough draft of the KEP ready. Will make the PR in a couple days

**Comment by krmayankk** [comment]
opened the KEP  https://github.com/kubernetes/enhancements/pull/678

**Comment by surajssd** [comment]
There is a similar issue https://github.com/kubernetes/kubernetes/issues/72320 where StatefulSet could not be evicted from a node due to PodDisruptionBudget has `maxUnavailable: 25%` set.

**Comment by iksaif** [comment]
FYI: I am interested in this for the exact same reasons :)

**Comment by krmayankk** [comment]
thanks @iksaif ,we have a followup proposal to soldify the semantics of this here https://github.com/kubernetes/enhancements/pull/1010

**Comment by krmayankk** [comment]
API is here https://github.com/kubernetes/kubernetes/pull/82162 , working on the implementation

**Comment by R-omk** [comment]
/remove-lifecycle stale

Moreover, I believe that he should respect PodDisruptionBudget, 
because now the update process may disrupt maxUnavailable  countrer.

**Comment by krmayankk** [comment]
i have started working back on this, will update the api and implementation pR

**Comment by krmayankk** [comment]
@kow3ns can you assign this to 1.20 milestone ?

**Comment by MarcPow** [comment]
@krmayankk Has this feature been slotted into a release?

**Comment by krmayankk** [comment]
@soltysh  for marking it for a release

**Comment by jyotimahapatra** [comment]
@krmayankk @soltysh I'm the bug triage shadow 👋 . We're in 1.23 code freeze. I wanted to check if this issue can be moved to milestone 1.24.

**Comment by krmayankk** [comment]
yes please @jyotimahapatra

**Comment by ritpanjw** [comment]
Hi 👋 . Checking in from bug triage for 1.24. Should we continue to track progress on this issue for 1.24 release?  If yes, would you please update the priority field for us to track it better?

**Comment by krmayankk** [comment]
@ritpanjw yes please, its complete, pending a few small changes to tests.

**Comment by ritpanjw** [comment]
@krmayankk Hi 👋 . Checking in from bug triage for 1.24, is the task complete or still small changes are pending ?

**Comment by ritpanjw** [comment]
Hi 👋 Im checking in from bug triage team for release 1.24, we are very near to code freeze, could you please update?

**Comment by dims** [comment]
@krmayankk let's take this off 1.24 as the code freeze is next week. tests can still merge and we can reinstate the v1.24 milestone if you are able to make progress on the tests. let's take the pressure of this on the release team.

/milestone v1.25

**Comment by krmayankk** [comment]
@dims  this is merged now https://github.com/kubernetes/kubernetes/pull/82162 , we can close this issue now

**Comment by nantoniazzi** [comment]
Why isn't it possible to set a value of 0?

In my scenario, I have 3 pods that are consuming a lot of resources. If I remove one pod during the rolling update, the other pods will then take all the load on them. I would prefer to over-provision my pod during the rolling update to be sure to always have enough resources available.

Wouldn't it be possible by setting `maxUnavailable` to 0? or is there another way of handling this scenario with a statefulset? `maxSurge` would be what I'm missing actually, but it only seems to be available for `deployments`. I need a statefulset because I have a lot of cached content on disk associated with my pods and I need a statefulset for that (`deployments` would add too much loading during startup without my cache)

**Comment by nantoniazzi** [comment]
> We discussed this in SIG Apps meeting today and we're looking for more requests from users and some more concrete use cases.
> 
> I'd also like to point out that MaxSurge does not make sense for StatefulSets -- you wouldn't want to create more replicas (surge) than a StatefulSet's `.spec.replicas` during a rolling update, as each pod has its own identity. MaxUnavailable is the only thing needed here.

I just realized that `maxSurge` was not to implemented on purpose with this comment. I disagree and there are scenarios where it's necessary. I do not think that's a problem to add more statefulset replicas during rolling update if it's the expected behavior of the cluster manager.

In our scenario, we need statefulset to have a dedicated cache for each pod (the cache is specific to each working pod), and we do not want to downscale them during a deployment because it will creates spike of loads on other pods.

What is the procedure now, should I open a feature request issue?

**Comment by hobbytp** [comment]
[Openkruise](https://openkruise.io/docs/user-manuals/advancedstatefulset/#maxunavailable) supports maxunavailable, so it sounds reasonable to support it.

**Comment by krmayankk** [comment]
@nantoniazzi yes we did not purposely implement that behavior. Please open an issue and we can discuss it more broadly with the community.

**Comment by hobbytp** [comment]
From https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.25/#rollingupdatestatefulsetstrategy-v1-apps, I found maxUnavailable is already there, so I think "maxUnavailable" is already supported in StatefulSet RollingUpdate.
There is also a blog https://kubernetes.io/blog/2022/05/27/maxunavailable-for-statefulset/, so seems k8s v1.24 supports it.
